src/tr1/game/savegame/savegame_bson.c

```diff
@@ -918,37 +918,52 @@ static bool M_LoadLara(
     return true;
 }
 
-static bool M_LoadCurrentMusic(JSON_OBJECT *music_obj)
+static bool M_LoadCurrentMusic(
+    JSON_OBJECT *music_obj, const uint16_t header_version)
 {
-    if (g_Config.audio.music_load_condition == MUSIC_LOAD_NEVER) {
-        return true;
-    }
-
-    if (!music_obj) {
+    if (music_obj == nullptr) {
         LOG_WARNING("Malformed save: invalid or missing current music");
         return true;
     }
 
-    int16_t current_track = JSON_ObjectGetInt(music_obj, "current_track", -1);
-    double timestamp = JSON_ObjectGetDouble(music_obj, "timestamp", -1.0);
-    if (current_track != MX_INACTIVE) {
-        const bool is_ambient =
+    const MUSIC_TRACK_ID current_track =
+        JSON_ObjectGetInt(music_obj, "current_track", MX_INACTIVE);
+    MUSIC_TRACK_ID ambient_track =
+        JSON_ObjectGetInt(music_obj, "current_ambient", MX_INACTIVE);
+    const double timestamp = JSON_ObjectGetDouble(music_obj, "timestamp", -1.0);
+
+    if (header_version < VERSION_9) {
+        const bool legacy_ambient =
             JSON_ObjectGetBool(music_obj, "is_ambient", false);
-        if (is_ambient) {
-            if (g_Config.audio.music_load_condition == MUSIC_LOAD_NON_AMBIENT) {
-                return true;
-            }
-            Music_Play(current_track, MPM_LOOPED);
-        } else {
-            Music_Play(current_track, MPM_ALWAYS);
+        if (legacy_ambient && current_track != MX_INACTIVE) {
+            ambient_track = current_track;
         }
+    }
 
-        if (!Music_SeekTimestamp(timestamp)) {
-            LOG_WARNING(
-                "Could not load current track %d at timestamp %" PRId64 ".",
-                current_track, timestamp);
-        }
+    if (ambient_track != MX_INACTIVE) {
+        // Always restart the ambient as it may have changed based on the
+        // current position in the level.
+        Music_Play(ambient_track, MPM_LOOPED);
+    }
+
+    if (g_Config.audio.music_load_condition == MUSIC_LOAD_NEVER) {
+        return true;
+    }
+
+    const bool is_ambient =
+        current_track != MX_INACTIVE && current_track == ambient_track;
+    if (!is_ambient && current_track != MX_INACTIVE) {
+        Music_Play(current_track, MPM_ALWAYS);
     }
+
+    const bool load_timestamp =
+        !is_ambient || g_Config.audio.music_load_condition == MUSIC_LOAD_ALWAYS;
+    if (load_timestamp && !Music_SeekTimestamp(timestamp)) {
+        LOG_WARNING(
+            "Could not load current track %d at timestamp %" PRId64 ".",
+            current_track, timestamp);
+    }
+
```

Comment:
Complex stuff but still coherent more or less. Wish we had tests for this

Reply:
Tests would be so good 🙏 One day.